File: player/CloudsAndDroids/app/src/main/java/ru/trinitydigital/cloudsanddroids/StatsFragment.java
package ru.trinitydigital.cloudsanddroids;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import butterknife.Bind;
import butterknife.ButterKnife;

public class StatsFragment extends Fragment {
    private static String TAG = "StatsFragment";
    private static final String PLAYER_ID = "PLAYER_ID";
    private static final String OPPONENT_ID = "OPPONENT_ID";
    private static final String BATTLE_ID = "BATTLE_ID";

    private String playerId;
    private String opponentId;
    private String battleId;
    FirebaseDatabase database;

    int maxHp;
    int maxMana;

    String hp;
    String mana;

    @Bind(R.id.noStats)
    View noStats;

    @Bind(R.id.someStats)
    View someStats;

    @Bind(R.id.name)
    TextView nameView;

    @Bind(R.id.hp)
    TextView hpView;

    @Bind(R.id.mana)
    TextView manaView;

    @Bind(R.id.opponent)
    ImageView opponentView;

    @Bind(R.id.opponentName)
    TextView opponentName;

    public StatsFragment() {
        // Required empty public constructor
    }

    public static StatsFragment newInstance(String battleId, String playerId, String opponentId) {
        StatsFragment fragment = new StatsFragment();
        Bundle args = new Bundle();
        args.putString(PLAYER_ID, playerId);
        args.putString(BATTLE_ID, battleId);
        args.putString(OPPONENT_ID, opponentId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            playerId = getArguments().getString(PLAYER_ID);
            battleId = getArguments().getString(BATTLE_ID);
            opponentId = getArguments().getString(OPPONENT_ID);
        }
        database = FirebaseDatabase.getInstance();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_stats, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);
        database.getReference().child("settings").addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot settings) {
                maxHp = Integer.parseInt(settings.child("max_hp").getValue().toString());
                maxMana = Integer.parseInt(settings.child("max_mana").getValue().toString());
                hpView.setText("HP: " + hp + "/" + maxHp);
                manaView.setText("MANA: " + mana + "/" + maxMana);
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                // Failed to read value
                Log.w(TAG, "Failed to read value.", databaseError.toException());
            }
        });

        database.getReference().child("battles").child(battleId).child("states").child(playerId)
                .addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot player) {

                        nameView.setText(player.child("name").getValue().toString());

                        if (player.hasChild("hp") && player.hasChild("mana")) {
                            someStats.setVisibility(View.VISIBLE);
                            noStats.setVisibility(View.GONE);

                            hp = player.child("hp").getValue().toString();
                            mana = player.child("mana").getValue().toString();

                            hpView.setText("HP: " + hp + "/" + maxHp);
                            manaView.setText("MANA: " + mana + "/" + maxMana);
                        }
                        else {
                            noStats.setVisibility(View.VISIBLE);
                            someStats.setVisibility(View.GONE);
                        }
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        // Failed to read value
                        Log.w(TAG, "Failed to read value.", databaseError.toException());
                    }
                });

        database.getReference().child("battles").child(battleId).child("states").child(opponentId)
                .addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot opponent) {
                        opponentName.setText(opponent.child("name").getValue().toString());

                        if (opponent.hasChild("hp") && opponent.hasChild("mana")) {
                            int hp = Integer.parseInt(opponent.child("hp").getValue().toString());
                            float thidPart = maxHp / 3.0f;
                            if (hp == 0) {
                                opponentView.setImageResource(R.drawable.grumpy);
                                return;
                            }
                            if (hp < thidPart) {
                                opponentView.setImageResource(R.drawable.sad);
                                return;
                            }
                            if (hp < thidPart * 2) {
                                opponentView.setImageResource(R.drawable.neutral);
                                return;
                            }
                            opponentView.setImageResource(R.drawable.smile);
                        }
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        // Failed to read value
                        Log.w(TAG, "Failed to read value.", databaseError.toException());
                    }
                });
    }
}

File: player/CloudsAndDroids/app/src/main/java/ru/trinitydigital/cloudsanddroids/StatsFragment.kt
package ru.trinitydigital.cloudsanddroids

import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.firebase.database.FirebaseDatabase

import kotlinx.android.synthetic.main.fragment_stats.*

class StatsFragment : Fragment() {

    internal var maxHp: Int = 0
    internal var maxMana: Int = 0

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_stats, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        FirebaseDatabase.getInstance()
                .reference
                .child("settings")
                .getValueSingle(Types.settingsType, Settings())
                .subscribe({
                    maxHp = it.max_hp
                    maxMana = it.max_mana
                    trackMyState()
                    trackOpponentState()
                    trackSpells()
                },{
                    // Failed to read value
                    Log.w(TAG, "Failed to read value.", it)
                })
    }

    fun trackMyState() {
        FirebaseDatabase.getInstance()
                .reference
                .child("battles")
                .child(arguments.getString(BATTLE_ID))
                .child("states")
                .child(arguments.getString(PLAYER_ID))
                .getValueObservable(Types.stateType, State())
                .subscribe({
                    name.text = it.name
                    hpText.text = "HP: ${it.hp}/$maxHp"
                    manaText.text = "MANA: ${it.mana}/$maxMana"
                },{
                    // Failed to read value
                    Log.w(TAG, "Failed to read value.", it)
                })
    }

    fun trackOpponentState() {
        FirebaseDatabase.getInstance()
                .reference
                .child("battles")
                .child(arguments.getString(BATTLE_ID))
                .child("states")
                .child(arguments.getString(OPPONENT_ID))
                .getValueObservable(Types.stateType, State())
                .subscribe({
                    opponentName.text = it.name
                    if (it.hp <= 0)
                        opponentImage.setImageResource(R.drawable.grumpy)
                    else if (it.hp < maxHp / 3.0f)
                        opponentImage.setImageResource(R.drawable.sad)
                    else if (it.hp < maxHp / 3.0f * 2)
                        opponentImage.setImageResource(R.drawable.neutral)
                    opponentImage.setImageResource(R.drawable.smile)
                },{
                    // Failed to read value
                    Log.w(TAG, "Failed to read value.", it)
                })
    }

    fun trackSpells() {
        FirebaseDatabase.getInstance()
                .reference
                .child("battles")
                .child(arguments.getString(BATTLE_ID))
                .child("turns")
                .orderByKey()
                .limitToLast(1)
                .getLastValueObservable(Types.turnType, Turn())
                .subscribe({
                    val target = it.target.toString()
                    if (!it.card.isNullOrEmpty())
                        FirebaseDatabase.getInstance()
                                .reference
                                .child("cards")
                                .child(it.card)
                                .getValueSingle(Types.cardType, Card())
                                .subscribe({
                                    spellText.text = it.name
                                    if (target == arguments.getString(OPPONENT_ID))
                                        spellText.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_trending_down_black_24dp, 0)
                                    else if (target == arguments.getString(PLAYER_ID))
                                        spellText.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_trending_up_black_24dp, 0)
                                },{
                                    // Failed to read value
                                    Log.w(TAG, "Failed to read value.", it)
                                })
                },{
                    // Failed to read value
                    Log.w(TAG, "Failed to read value.", it)
                })
    }

    companion object {
        private val TAG = "StatsFragment"
        private val PLAYER_ID = "PLAYER_ID"
        private val OPPONENT_ID = "OPPONENT_ID"
        private val BATTLE_ID = "BATTLE_ID"

        fun newInstance(battleId: String, playerId: String, opponentId: String): StatsFragment {
            val fragment = StatsFragment()
            val args = Bundle()
            args.putString(PLAYER_ID, playerId)
            args.putString(BATTLE_ID, battleId)
            args.putString(OPPONENT_ID, opponentId)
            fragment.arguments = args
            return fragment
        }
    }
}
